File: 662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    class Pair{
        int x;
        TreeNode node;
        Pair(TreeNode node,int x){
            this.x=x;
            this.node=node;
        }
    }
    public int widthOfBinaryTree(TreeNode root) {
        Queue<Pair> q=new LinkedList<>();
        int ans=0;

        q.add(new Pair(root,1));
        while(!q.isEmpty()){
            int s=q.size();
            int f=0,l=0;
            for(int i=0;i<s;i++){
                Pair p=q.remove();
                if(i==0) f=p.x;
                if(i==s-1) l=p.x;

                if(p.node.left!=null) q.add(new Pair(p.node.left,2*p.x-f));
                if(p.node.right!=null) q.add(new Pair(p.node.right,2*p.x+1-f));
            }
            ans=Math.max(ans,l-f+1);
        }

        return ans;
    }
}

File: 662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def widthOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root == None:
            return 0

        q = [(root, 1)]
        ans = 0

        while q:
            s = len(q)
            f = -1
            e = -1

            for i in range(s):
                node, c = q.pop(0)
                if i == 0:
                    f = c
                if i == s-1:
                    e = c
                
                if node.left != None:
                    q.append((node.left, 2 * c + 1 - f))
                if node.right != None:
                    q.append((node.right, 2 * c + 2 - f))

            ans = max(ans, e - f + 1)
        
        return ans
                    



        